src/clients/rust/src/time_based_id.rs

```diff
@@ -0,0 +1,314 @@
+                        // "millisecond" could last forever.
+                        //
+                        // The ULID spec says if the randomness overflows 80
+                        // bits then it is an error, but we instead manually
+                        // carry the bit from the randomness to the timestamp,
+                        // moving time forward by 1ms, and generate a new random
+                        // number. Generated ids will be ahead of the wall
+                        // clock by 1 ms until time catches up.
+                        TbidGenerator {
+                            // There is too much resolution in a u128 for this to ever overflow.
+                            ms_since_epoch: previous_ms_since_epoch
+                                .checked_add(1)
+                                .expect("impossible overflow"),
+                            random: random_u80(),
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn next_random(&self) -> Option<u128> {
+        assert!(is_u80(self.random));
+        if self.random == U80_MASK {
+            None
+        } else {
+            Some(self.random + 1)
+        }
+    }
+}
+
+const U80_MASK: u128 = 0x_FFFF_FFFF_FFFF_FFFF_FFFF;
+
+fn is_u80(val: u128) -> bool {
+    val <= U80_MASK
+}
+
+fn random_u80() -> u128 {
+    random_u128() & U80_MASK
+}
+
+fn random_u128() -> u128 {
+    let a = random_u64();
+    let b = random_u64();
+    let a = a as u128;
+    let b = b as u128;
+    a | b << 64
+}
+
+fn random_u64() -> u64 {
+    std::hash::Hasher::finish(&std::hash::BuildHasher::build_hasher(
+        &std::collections::hash_map::RandomState::new(),
+    ))
+}
+
+#[cfg(test)]
+use std::time::{Duration, Instant};
+
+#[cfg(test)]
+struct TestResult {
+    id_start: u128,
+    id_end: u128,
+}
+
+#[cfg(test)]
+fn run_test_for_a_few_ms(mut id_generator: impl FnMut() -> u128) -> TestResult {
+    // Enough time to transition between millisecond timestamps
+    let test_duration = Duration::from_millis(10);
+    let end_instant = Instant::now() + test_duration;
+
+    let mut id_prev = 0;
+    let mut id_start = None;
+    let mut id_end = None;
+
+    while Instant::now() < end_instant {
+        let id_next = id_generator();
+
+        assert!(id_prev < id_next, "ids not monotonically increasing");
+
+        // Regression test - should not reset random to 0 on overflow,
+        // but instead generate a new random number.
+        let id_next_random_part = id_next & U80_MASK;
+        assert_ne!(id_next_random_part, 0);
+
+        // Don't waste other processes' time
+        std::thread::yield_now();
+
+        id_start = id_start.or(Some(id_next));
+        id_end = Some(id_next);
+
+        id_prev = id_next
+    }
+
+    let id_start = id_start.unwrap();
+    let id_end = id_end.unwrap();
+
+    TestResult { id_start, id_end }
+}
+
+#[cfg(test)]
```

Comment:
Slight preference for defining 

```
#[cfg(test)]
mod tests {
  fn helper() { ... }

  #[test]
  fn test() { ... }
}
```

here, to avoid individual cfg gates, and to make it abundantly clear which code relates to tests and which is production.